bftengine/src/bcstatetransfer/Messages.hpp

```diff
@@ -40,55 +40,52 @@ class MsgType {
 };
 
 struct BCStateTranBaseMsg {
+  BCStateTranBaseMsg(uint16_t type) : type(type) {}
   uint16_t type;
+  // This struct and its derived structs are used to de/serialize message buffers sent over communication channels.
+  // Since virtual methods modify the memory layout of a struct, there cannot be any for both this base struct and its
+  // inheritors.
+};
+
+// VariableSizeMsg is useful for structs with a flexible array member
+template <typename T>
+class VariableSizeMsg {
+ public:
+  VariableSizeMsg(size_t dataSize) : buffer_(new concord::Byte[calcMsgSize(dataSize)]()) {}
+
+  T* operator->() const { return reinterpret_cast<T*>(buffer_.get()); }
+
+  char* getSerializedMsg() const { return reinterpret_cast<char*>(buffer_.get()); }
```

Comment:
This is still a bad idea not to return a byte type, especially when the only usage for this method is in tests.
If the casts hurt you so much, you can make it a template function so that it can receive both a char* and concord::Byte.
This is not the best approach as it is still vulnerable to programmers mistaking bytes with text characters, but at least this way the intent is bit clearer. I also still think that the name of this method should be getBytes, or if you want to be more explicit (and have redundancy in the name of the function) getMsgBytes.